### PR TITLE
Part 5

### PR DESCRIPTION
**Pull request by Jag50**
Boundary cases are tested
Strikes and spares in frame 10 with bonus balls tested
Test and implement input validation complete

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -42,6 +42,30 @@
             <version>1.3</version>
             <scope>test</scope>
         </dependency>
+        <dependency>
```
Comment (rachelwarburton) [inline pom.xml:45]: There are two versions for the junit-jupiter-api?

Comment (Jag50) [reply pom.xml:45]: I added additional dependancies while trying out various methods

File: src/main/java/bowlingexercise3/Bowling.java [hunk 2]
```diff
@@ -34,6 +57,19 @@ public void turns(String total) {
 
     }
 
+    private void validateInput(String input) throws IllegalArgumentException {
+
+        //should have 10 frames
+        // List<String> frames = Arrays.asList(input.split("\\|"));
```
Comment (rachelwarburton) [inline src/main/java/bowlingexercise3/Bowling.java]: Code that's no longer in use can be deleted.

Comment (Jag50) [reply src/main/java/bowlingexercise3/Bowling.java]: Addressed.

**Review by rachelwarburton — COMMENTED** [review]
I know the test coverage is good, but I think it would help to tidy up the tests, thinking about the required coverage and giving clear names as to what each group of tests is doing, so test methods have names like invalidNofFramesThrowsException etc so we can see at a glance the types of input validation that are covered. That way when you get a test failure after doing some refactoring, it's also clear when tests start to fail, what part of the solution is no longer working.

File: src/test/java/bowlingexercise3/BowlingTest.java [hunk 3]
```diff
@@ -61,22 +81,339 @@ public void testBowlingSpares(String turns, int score) {
...
+            "--|--|--|--|--|--|--|--|--|-/||4",
+            "--|--|--|--|--|--|--|--|--|-/||6",
+            "--|--|--|--|--|--|--|--|--|-/||7",
+            "--|--|--|--|--|--|--|--|--|-/||8",
+            "--|--|--|--|--|--|--|--|--|-/||9",
+            "--|--|--|--|--|--|--|--|--|-/||X",
+            "--|--|--|--|--|--|--|--|--|-/||5",
+    })
+
+    public void boundaryConditionsSpareBonusBallTest(String turns) {
```
Comment (rachelwarburton) [inline src/test/java/bowlingexercise3/BowlingTest.java]: These are input validation tests for 10th frame containing a spare?

File: src/test/java/bowlingexercise3/BowlingTest.java [hunk 4]
```diff
@@ -61,22 +81,339 @@ public void testBowlingSpares(String turns, int score) {
...
+            "X|X|X|X|X|X|X|X|X|?||35",
+            "X|X|X|X|X|X|X|X|X|X||81-",
+            "9|X|X|X|X|X|2X|X|X|X||",
+            "9|X|X|X|X|X|X|X|X|X||",
+            "/X|X|X|X|X|X|X|X|X|X||",
+            "-/|X|X|X|X|X|X|X|X|X||-X",
+
+
+    })
+    public void TestingInvalidInput(String turns) {
```
Comment (rachelwarburton) [inline src/test/java/bowlingexercise3/BowlingTest.java]: I get confused between this method and the previous one? So there are some tests for the wrong number of frames, invalid characters in frames 1-9, frame 10 and bonus balls?

File: src/test/java/bowlingexercise3/BowlingTest.java [hunk 5]
```diff
@@ -61,22 +81,339 @@ public void testBowlingSpares(String turns, int score) {
...
+            "5/|5/|5/|5/|5/|5/|5/|5/|5/|5/||",
+            "6/|6/|6/|6/|6/|6/|6/|6/|6/|6/||",
+            "7/|7/|7/|7/|7/|7/|7/|7/|7/|7/||",
+            "8/|8/|8/|8/|8/|8/|8/|8/|8/|8/||",
+            "9/|9/|9/|9/|9/|9/|9/|9/|9/|9/||",
+
+
+
+    })
+    public void testingValidInputNormalFrames(String turns) {
```
Comment (rachelwarburton) [inline src/test/java/bowlingexercise3/BowlingTest.java]: normal - is that frames without strikes or spares?

File: src/test/java/bowlingexercise3/BowlingTest.java [hunk 6]
```diff
@@ -61,22 +81,339 @@ public void testBowlingSpares(String turns, int score) {
...
+            "X|11|11|11|11|11|11|11|11|11||",
+            "X|11|X|11|X|11|X|11|X|11||",
+            "X|11|11|11|X|X|X|X|X|X||",
+            "X|12|23|X|X|X|9-|X|X|X||",
+            "X|11|11|11|11|11|11|11|X|X||",
+
+
+
+    })
+    public void testValidInputStrikeNormal10Frames(String turns) {
```
Comment (rachelwarburton) [inline src/test/java/bowlingexercise3/BowlingTest.java:318]: Is it testing strikes in all frames 1-10 - are these input strings valid? Some of them end in strikes, but don't have any bonus balls?

File: src/test/java/bowlingexercise3/BowlingTest.java [hunk 7]
```diff
@@ -61,22 +81,339 @@ public void testBowlingSpares(String turns, int score) {
...
+
+            "X|-/|X|-/|X|-/|X|-/|X|-/||",
+            "12|-/|X|-/|X|-/|X|-/|X|-/||",
+            "X|34|X|-/|X|-/|X|-/|X|-/||",
+            "X|-/|54|-/|X|-/|X|-/|X|-/||",
+            "X|-/|1/|-/|X|-/|X|-/|X|-/||",
+            "X|-/|X|-/|X|9/|X|-/|X|-/||",
+
+
+    })
```
Comment (rachelwarburton) [inline src/test/java/bowlingexercise3/BowlingTest.java:335]: I am wondering why these inputs are valid as there are no bonus balls?

File: src/test/java/bowlingexercise3/BowlingTest.java [hunk 8]
```diff
@@ -61,22 +81,339 @@ public void testBowlingSpares(String turns, int score) {
...
+            "5/|5/|5/|5/|5/|5/|5/|5/|5/|5/||+|",
+            "6/|6/|6/|6/|6/|6/|6/|6/|6/|6/|||9",
+            "7/|7/|7/|7/|7/|7/|7/|7/|7/|7/||/",
+            "8/|8/|8/|8/|8/|8/|8/|8/|8/|8/||£",
+            "9/|9/|9/|9/|9/|9/|9/|9/|9/|9/||:",
+            "1/|1/|3/|4/|5/|6/|7/|7/|8/|9/||X4",
+
+
+    })
+    public void testInvalidInputBonusFrames(String turns) {
```
Comment (rachelwarburton) [inline src/test/java/bowlingexercise3/BowlingTest.java:413]: Maybe this could be split into two groups for invalid bonus strikes and invalid bonus spares? Though I can see why only one test is necessary, it could make it clearer for regression purposes as we know there are two different supporting regular expressions implementing those cases.

**Review by rachelwarburton — COMMENTED** [review]
It is getting much clearer thank you - it is worth making the tests clear as when you've been immersed in something it is clear, but when you come back to something a week or more later and want to do some refactoring, you have a really clear set of test cases that are maintainable.

File: src/test/java/bowlingexercise3/BowlingTest.java (hunk 3 above)
Comment (Jag50) [reply src/test/java/bowlingexercise3/BowlingTest.java]: These are input validations for a bonus on the 10th frame which is a spare.

File: src/test/java/bowlingexercise3/BowlingTest.java (hunk 4 above)
Comment (Jag50) [reply src/test/java/bowlingexercise3/BowlingTest.java]: Yes, I have separated tests into frame sizes, with/without bonus frames, and other variations

Comment (Jag50) [reply src/test/java/bowlingexercise3/BowlingTest.java]: In this instance, I merged them into one humoungous invalid test

File: src/test/java/bowlingexercise3/BowlingTest.java (hunk 5 above)
Comment (Jag50) [reply src/test/java/bowlingexercise3/BowlingTest.java]: Yes, Normal implies no bonus ball

File: src/test/java/bowlingexercise3/BowlingTest.java (hunk 6 above)
Comment (Jag50) [reply src/test/java/bowlingexercise3/BowlingTest.java:318]: I wanted to check whether strikes in any position work and are recognised as valid before I moved onto the bonus ball criteria.

File: src/test/java/bowlingexercise3/BowlingTest.java (hunk 3 above)
Comment (rachelwarburton) [reply src/test/java/bowlingexercise3/BowlingTest.java]: Would it be called something like assertValidInputForSpareBonusBalls()? What are the boundary conditions?

File: src/test/java/bowlingexercise3/BowlingTest.java (hunk 4 above)
Comment (rachelwarburton) [reply src/test/java/bowlingexercise3/BowlingTest.java]: Could the test methods have different names or be split into groups? It's hard to tell what their purpose is as there are two methods with the same names?

File: src/test/java/bowlingexercise3/BowlingTest.java (hunk 3 above)
Comment (Jag50) [reply src/test/java/bowlingexercise3/BowlingTest.java]: Addressed and renamed appropriately

File: src/test/java/bowlingexercise3/BowlingTest.java (hunk 6 above)
Comment (rachelwarburton) [reply src/test/java/bowlingexercise3/BowlingTest.java:318]: ok, but I thought it would throw an IllegalArgumentException if there were no bonus balls following a strike in the 10th frame?

File: src/test/java/bowlingexercise3/BowlingTest.java (hunk 4 above)
Comment (Jag50) [reply src/test/java/bowlingexercise3/BowlingTest.java]: Have labelled it more clearly now

File: src/test/java/bowlingexercise3/BowlingTest.java (hunk 7 above)
Comment (Jag50) [reply src/test/java/bowlingexercise3/BowlingTest.java:335]: Updated

File: src/test/java/bowlingexercise3/BowlingTest.java (hunk 6 above)
Comment (Jag50) [reply src/test/java/bowlingexercise3/BowlingTest.java:318]: Updated

**Review by rachelwarburton — APPROVED** [review]
looks fine to me - we spotted a potential defect where a strike in the 10th frame was valid without bonus balls, but can refactor later if there is time.
